tensorflow-utils/src/main/java/org/tensorflow/util/ndarray/NdArray.java

```diff
@@ -0,0 +1,356 @@
+   * an exception is thrown. After the copy, the content of both arrays can be altered
+   * independently, without affecting each other.
+   *
+   * @param dst array to receive a copy of the content of this array
+   * @return this array
+   * @throws IllegalArgumentException if the shape of {@code dst} is not equal to the shape of this
+   * array
+   */
+  NdArray<T> copyTo(NdArray<T> dst);
+
+  /**
+   * Read the content of this N-dimensional array into the destination buffer.
+   *
+   * <p>The size of the buffer must be equal or greater to the {@link #size()} of this
+   * array, or an exception is thrown. After the copy, content of the buffer and of the array can be
+   * altered independently, without affecting each other.
+   *
+   * @param dst the destination buffer
+   * @return this array
+   * @throws java.nio.BufferOverflowException if the buffer cannot hold the content of this array
+   * @see DataBuffer#size()
+   */
+  NdArray<T> read(DataBuffer<T> dst);
+
+  /**
+   * Write the content of this N-dimensional array from the source buffer.
+   *
+   * <p>The size of the buffer must be equal or greater to the {@link #size()} of this
+   * array, or an exception is thrown. After the copy, content of the buffer and of the array can be
+   * altered independently, without affecting each other.
+   *
+   * @param src the source buffer
+   * @return this array
+   * @throws java.nio.BufferUnderflowException if the buffer has not enough remaining data to write
+   * into this array
+   * @see DataBuffer#size()
+   */
+  NdArray<T> write(DataBuffer<T> src);
+
+  /**
+   * Reads the content of this N-dimensional array into the destination array.
+   *
+   * <p>The size of the destination array must be equal or greater to the {@link #size()} of this
+   * array, or an exception is thrown. After the copy, content of the both arrays can be altered
+   * independently, without affecting each other.
+   *
+   * @param dst the destination array
+   * @return this array
+   * @throws java.nio.BufferOverflowException if the destination array cannot hold the content of
+   * this array
+   */
+  NdArray<T> read(T[] dst);
+
+  /**
+   * Reads the content of this N-dimensional array into the destination array.
+   *
+   * <p>{@code dst.length - offset} must be equal or greater to the {@link #size()} of this array,
+   * or an exception is thrown. After the copy, content of the both arrays can be altered
+   * independently, without affecting each other.
+   *
+   * @param dst the destination array
+   * @param offset the index of the first element to write in the destination array
+   * @return this array
+   * @throws java.nio.BufferOverflowException if the destination array cannot hold the content of
+   * this array
+   * @throws IllegalArgumentException if offset is greater than dst length or is negative
+   */
+  NdArray<T> read(T[] dst, int offset);
```

Comment:
Shouldn't `offset` be a long? And similarly for write. I think this method is rather confusing as to know the offset you need to know what the backing store is and if it's row or column major etc. I also couldn't easily find a way of calculating the offset.

Reply:
That `offset` is the offset in the array passed in argument (therefore 32-bits only). That is to mimics methods found in `java.nio.Buffer`, I don't know how useful they are but it was not hard to support. I admit though that the double usage of `offset` might be confusing... maybe I can rename it to `dstOffset`? (and `srcOffset` for write)?

Reply:
Ah, ok. So it starts reading/writing from the start of the `NdArray` and fills it in whatever order the backing store uses? I still think that we should document the orientation of the backing store (i.e. row or column major) and if we want it to be adjustable (e.g. to allow passing matrices to fortran code) then it should be an enum stored in an instance field and available on all the `NdArray` subtypes.

Reply:
You mean the enum should be in the `NdArray` to tell what is the orientation of the data in its buffer or it should be provided by the caller to ask `NdArray` to read/write its data to the src/dest array in a specific order?

Reply:
I think both, as if they differ then the call will be very slow, so users need to be able to see and control the orientation. For example, a row wise view is important when you are doing SGD, but a column wise view is useful for doing feature selection, coordinate ascent (e.g. in an ElasticNet) or building a decision tree.

Reply:
I think it is a good idea but changing the orientation of an array may have a significant impact on the actual solution. What can be done meanwhile is simply to permute dimensions of your tensor using TF before performing the copy (e.g. transposing a 2d matrix).

Reply:
I agree in TF I have other ways of doing it, but not within the ndarray interface itself, and it's a property of the ndarray not just an ndarray as applied to TF. I think it needs to allow the user some control over the layout, but it doesn't have to be in this initial PR.